mlsim/bias_components.py

```diff
@@ -0,0 +1,449 @@
+        z = [np.random.choice([0,1], p= self.params.Pz_a[ai]) for ai in a]
+
+        return np.asarray(a).T,np.asarray(z).T
+
+    def get_rho_a(self):
+        '''
+        get  P(A=1)
+
+        Parameters
+        -----------
+
+        Returns
+        -------
+        rho_a : float
+            Probability of being in the disadvantaged group, A =1
+        '''
+        return self.params.Pa[1]
+
+    def get_rho_z(self):
+        '''
+        return P(Z=1|A)
+
+        Parameters
+        -----------
+
+        Returns
+        -------
+        rho_z : nparray of floats
+            probability of the favorable outcome(z =1) for A=0 and A=1 in that
+            order
+        '''
+
+        return np.asarray(self.params.Pz_a)[:,1]
+
+class DemographicIndependent(Demographic):
+    '''
+    '''
+    def __init__(self,rho_a=.2,rho_z=.1):
+        '''
+        P(A = 1) = rho_a
+        P(Z=1) = rho_z
+
+        default is independent sampling of a and z
+        '''
+        super().__init__(rho_a,rho_z)
+
+
+class DemographicCorrelated(Demographic):
+    '''
+    '''
+
+    def __init__(self,rho_a=.5,rho_z=[.5,.3]):
+        '''
+        P(A = 1) = rho_a
+        P(Z=1,A=0) = rho_z[0]
+        P(Z=1,A=1) = rho_z[1]
+
+        '''
+        Pa = [1-rho_a, rho_a]
+        Pz_a = [[1-rho_z[0], rho_z[0]],
+             [1-rho_z[1], rho_z[1]]]
+        Sampler.__init__(self,(Pa,Pz_a))
+
+
+class Target(Sampler):
+    '''
+    '''
+    ParamCreator = TargetParams
+    def __init__(self,beta=0.05):
+        '''
+        P(Y=Z|A,Z ) = P(Y=Z) = 1-beta
+        make errors with prob beta
+
+        beta =0, makes Y =Z
+        '''
+        pyeqz = [1-beta,beta]
+        Py_az = [[pyeqz,pyeqz],[pyeqz,pyeqz]]
+        super().__init__((Py_az,))
+
+
+    def sample(self,a,z):
+        '''
+        sample P(Y|A,Z) via P(Y=Z|A,Z)
+        Parameters
+        -----------
+        a :
+        z :
+        beta : float
+
+        '''
+
+        y = [np.random.choice([zi,1-zi],p= self.params.Py_az[ai][zi])
+                                            for ai,zi in zip(a,z)]
+
+        return np.asarray(y).T
+
+
+class TargetDisadvantagedError(Target):
+    '''
+    '''
+    def __init__(self,beta=.1):
+        '''
+        P(Y=Z|A=1,Z ) = P(Y=Z|A=1) = 1-beta
+        P(Y=Z|A=0,Z ) = P(Y=Z|A=0) = 1
+        make errors with prob beta
+
+        '''
+        pyeqz = [1-beta,beta]
+        Py_az = [[[1,0],[1,0]],[pyeqz,pyeqz]]
+        Sampler.__init__((Py_az,))
+
+class TargetTwoError(Target):
+    '''
+    '''
+    def __init__(self,beta=[0,.1]):
+        '''
+        P(Y=Z|A=1,Z ) = P(Y=Z|A=1) = 1-beta
+        P(Y=Z|A=0,Z ) = P(Y=Z|A=0) = 1
+        make errors with prob beta
+
+        '''
+        pyz_a0 = [1-beta[0],beta[0]]
+        pyz_a1 = [1-beta[1],beta[1]]
+        Py_az = [[pyz_a0,pyz_a0],[pyz_a1,pyz_a1]]
+        Sampler.__init__((Py_az,))
+
+mean_only_mvn = lambda mu :np.random.multivariate_normal(mu,np.eye(len(mu)))
```

Comment:
can this be moved inside the Feature class?

Reply:
like how? it's the default value, so if you don't pass a function for mean it uses that, but otherwise you passa sampling function